File: src/main/java/dev/zenhao/melon/gui/clickgui/component/SettingButton.java
package dev.zenhao.melon.gui.clickgui.component;

import dev.zenhao.melon.gui.clickgui.component.Component;
import dev.zenhao.melon.setting.ModeSetting;
import dev.zenhao.melon.setting.Setting;

public abstract class SettingButton<T>
extends Component {
    private Setting<T> value;

    public Setting<T> getValue() {
        return this.value;
    }

    public ModeSetting getAsModeValue() {
        return (ModeSetting)this.value;
    }

    public void setValue(Setting<T> value) {
        this.value = value;
    }
}

File: src/main/kotlin/melon/system/render/component/impl/settings/SettingButton.kt
package melon.system.render.component.impl.settings

import melon.system.config.setting.type.BooleanSetting
import melon.system.render.component.impl.basic.BooleanSlider
import dev.zenhao.melon.utils.vector.Vec2f

class SettingButton(val setting: BooleanSetting) : BooleanSlider(setting.name, setting.description, setting.visibility) {
    override val progress: Float
        get() = if (setting.value) 1.0f else 0.0f

    override fun onClick(mousePos: Vec2f, buttonId: Int) {
        super.onClick(mousePos, buttonId)
        setting.value = !setting.value
    }
}